(470) First cut of the Stock reports

parisiale opened the pull request:
Adding functionality for creating and exporting;
 - stock summary report;
 - stock movement report.

parisiale committed:
Add README.md

parisiale committed:
First cut of the Stock Export functionality

crwr45 commented on models/stock_export.py:
```diff
@@ -0,0 +1,282 @@
+        # locations like 'TRW/Stock' (for which the name is actually
+        # 'Stock', ofc...); ensure we want to also specify it
+        return self.env['stock.location'].search([('name', '=', 'Stock'),
+                                                  ('location_id', '=', 'UPL')])
```
location_id should be matching an integer, shouldn't it? By my understanding this shouldn't work.

parisiale replied:
Oh, mistake, I've tried this with `odoo-bin shell` by using the `name` attribute.

parisiale replied:
I think we can simply remove the parent check.

crwr45 commented on models/stock_export.py:
```diff
@@ -0,0 +1,282 @@
+        get_pallets_by_loc = lambda quant_union: {
+            Location.browse(loc): Quant.union(*_quants).mapped('package_id')
+            for loc, _quants in groupby(quant_union.sorted(key=by_location),
+                                        key=by_location)}
```
Looks like this lambda is getting the package the quant is in, which is not necessarily a pallet. 
We are removing 'pallet' from the terminology and using 'package' and 'parent package'. In this case I think it should be 'package'.

crwr45 commented on models/stock_export.py:
```diff
@@ -0,0 +1,282 @@
+            contained_quants = Quant.search(
+                [('package_id', 'child_of', pallet.ids)])
+            prod_quants = contained_quants.filtered(
+                lambda x: x.product_id == prod)
```
The filter should be moved to be part of the initial search domain.

crwr45 commented on models/stock_export.py:
```diff
@@ -0,0 +1,282 @@
+
+        Move = self.env['stock.move']
+        picking_type_in = self.env.ref('stock.picking_type_in')
+        picking_type_out = self.env.ref('stock.picking_type_out')
```
These picking types need to be accessed via the logged in users current warehouse (Users.get_user_warehouse())

parisiale replied:
Thanks, will change.

crwr45 commented on models/stock_export.py:
```diff
@@ -0,0 +1,282 @@
+
+        for goods_moves, sheet_name in [(in_moves, "Goods In"),
+                                        (out_moves, "Goods Out")]:
+            sheet = wb.add_sheet(sheet_name)
```
The method above used to create a sheet should be used here.

parisiale replied:
Yeah, but the above one is an inner closure, around the `Workbook` instance, so, if you make it an actual method, you should add that to the args. Besides that, another reason why I did not refactor is that I'm no sure of the impact of adding methods to model classes (and btw, is it necessary to add the `@api.model` decorator?).

crwr45 commented on models/stock_export.py:
```diff
@@ -0,0 +1,282 @@
+            attachment=attachment)
+
+
+# @todo: (ale) consider moving this helper elsewhere
```
Definitely move it elsewhere. I would suggest either adding to to the res.users model, or the mail.message model.

crwr45 reviewed (commented):
The references to 'pallet' should be genericised to 'package'.

The code structure is also not sitting right with me. There are two different reports exported by the same class, shown to the user as two different menu items. I think it would make sense to have the data extraction and file creation code on seperate models for the two reports.

parisiale commented:
@crwr45 thanks!

The code structure follows prior art.
I'm not sure I understand the benefit of what you're suggesting with the two separate models for data / file. The file-related logic (determine the file name and process a `xlwt` workbook) is tied to the fact we're using the excel format. And `xlwt` is something you use in the method that retrieves the data.

parisiale committed:
Remove parent clause for default inc. locations

Removing the check for the parent of the default 'Stock'
location; note that such check was broken, as it was not
performed against the actual 'name' of the parent location.

parisiale committed:
Remove todo messages

parisiale committed:
Use 'package' instead of 'pallet'

Updating naming.

parisiale committed:
Move helper func to inherited res.users

parisiale committed:
Remove superfluous filter

Moving filtering to the actual search query when
determining product quantities for Stock Summary.

parisiale committed:
Refactor helper fuction for creating spreadsheet tabs

parisiale committed:
Rename var

parisiale committed:
Use current warehouse for accessing picking type
